File: reordered-power-of-2/main.go
package main

func reorderedPowerOf2(n int) bool {
	digits := 0
	t := n
	for t > 0 {
		digits++
		t /= 10
	}

	lower := 0
	upper := 0
	for i := 0; i < digits; i++ {
		if i == 0 {
			lower = 1
			upper = 9
		} else {
			lower *= 10
			upper = upper*10 + 9
		}
	}

	candidates := []int{}
	start := 1
	for start <= upper {
		if start >= lower {
			candidates = append(candidates, start)
		}
		start *= 2
	}

	for _, candidate := range candidates {
		if check(candidate, n) {
			return true
		}
	}

	return false
}

func check(candidate, n int) bool {
	memory := make(map[int]int)
	for n > 0 {
		memory[n%10]++
		n /= 10
	}

	for candidate > 0 {
		v := candidate % 10
		if memory[v] <= 0 {
			return false
		}
		memory[v]--
		candidate /= 10
	}

	return true
}

File: reordered-power-of-2/main.py
class Solution:
    def reorderedPowerOf2(self, n: int) -> bool:
        nd = self.digits(n)
        p = 1
        d = self.digits(p)
        candidates = []
        while d <= nd:
            if d == nd:
                candidates.append(p)
            p *= 2
            d = self.digits(p)

        for c in candidates:
            if self.check(n, c):
                return True

        return False           

    def check(self, n, c):
        memo1 = defaultdict(int)
        while n > 0:
            memo1[n%10] += 1
            n //= 10

        memo2 = defaultdict(int)
        while c > 0:
            memo2[c%10] += 1
            c //= 10

        return memo1 == memo2

    def digits(self, n):
        r = 0
        while n > 0:
            r += 1
            n //= 10

        return r
